scripts: west_commands: patch: add src module argument

MeisterBob opened the pull request:
This patch adds the `--src-module` argument to west patch to be able to select the module where west patch looks for the patch definition without having to set `--patch-yml` and `--patch-base`. For clarification the present `--module` argument is renamed to `--dest-module`.

This patch also changes `--dest-module` and `module:` in `patches.yml` to take the name of the module instead of the path.

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -190,12 +199,20 @@ def filter_args(self, args):
             args.patch_base = manifest_dir / args.patch_base.relative_to(_WEST_MANIFEST_DIR)
-        if args.patch_yml.is_relative_to(_WEST_MANIFEST_DIR):
+        else:
+            args.patch_base = manifest_dir / args.patch_base
```
I don't like this approach of depending on the argument's type to change the behavior.

It's currently expanded (by default) to `Path("WEST_MANIFEST_DIR/zephyr/patches")`, so the relative aspect can still apply.

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -190,12 +207,35 @@ def filter_args(self, args):
         topdir = Path(self.topdir)
         manifest_dir = topdir / manifest_path
```
Wouldn't it be sufficient to change the `manifest_dir` based on some argument?

MeisterBob commented:
Thank you for your review @pdgendt. I have incorporated your suggestions and I like the code much better now too.

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -188,14 +205,28 @@ def filter_args(self, args):
 
         if args.patch_base.is_relative_to(_WEST_MANIFEST_DIR):
             args.patch_base = manifest_dir / args.patch_base.relative_to(_WEST_MANIFEST_DIR)
+        else:
```
```suggestion
        elif not args.patch_base.is_absolute():
```

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -188,14 +205,28 @@ def filter_args(self, args):
+
         if args.patch_yml.is_relative_to(_WEST_MANIFEST_DIR):
             args.patch_yml = manifest_dir / args.patch_yml.relative_to(_WEST_MANIFEST_DIR)
+        else:
```
```suggestion
        elif not args.patch_yml.is_absolute():
```

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -188,14 +205,28 @@ def filter_args(self, args):
+
         if args.west_workspace.is_relative_to(_WEST_TOPDIR):
             args.west_workspace = topdir / args.west_workspace.relative_to(_WEST_TOPDIR)
+        else:
```
```suggestion
        elif not args.west_workspace.is_absolute():
```

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -188,14 +205,28 @@ def filter_args(self, args):
+            mod = self.get_module(args.src_module)
+            if mod is None:
+                self.die("src-module not found")
+            manifest_dir = Path(mod.project)
```
```suggestion
            manifest_dir = topdir / mod.project
```

pdgendt commented on scripts/west_commands/patch.py:
```diff
@@ -188,14 +205,28 @@ def filter_args(self, args):
+        if args.src_module is not None:
+            mod = self.get_module(args.src_module)
+            if mod is None:
+                self.die("src-module not found")
```
```suggestion
                self.die(f'Source module "{args.src_module}" not found')
```

pdgendt reviewed (commented):
Looks reasonable, some suggestions.

cfriedt commented on scripts/west_commands/patch.py:
```diff
@@ -93,15 +99,17 @@ def do_add_parser(self, parser_adder):
             "-l",
             "--patch-yml",
-            help="Path to patches.yml file",
+            help=f"Path to patches.yml file (relative to module dir, default: {WEST_PATCH_YAML})",
```
Technically, this doesn't need to be relative, but the default probably should be

cfriedt commented on scripts/west_commands/patch.py:
```diff
@@ -115,16 +123,25 @@ def do_add_parser(self, parser_adder):
+        )
+        parser.add_argument(
+            "-dm",
+            "--dest-module",
```
Just a nit: why not use `dst` if using `src`?

cfriedt commented on scripts/west_commands/patch.py:
```diff
@@ -115,16 +123,25 @@ def do_add_parser(self, parser_adder):
-            "Option can be passed multiple times. "
-            "If this option is not given, the 'patch' command will run for Zephyr "
-            "and all modules.",
+            dest="dest_modules",
```
```suggestion
            dest="dst_modules",
```

cfriedt commented on scripts/west_commands/patch.py:
```diff
@@ -338,26 +379,23 @@ def clean(self, args, yml, mods=None):
-            mod_path = os.path.realpath(Path(args.west_workspace) / mod)
-            mod_paths[mod] = mod_path
+    def get_module(self, module):
+        all_modules = zephyr_module.parse_modules(ZEPHYR_BASE, self.manifest)
+        if module == 'zephyr':
+            Module = namedtuple('Module', ['project'])
+            return Module(ZEPHYR_BASE)
+        for m in all_modules:
+            if m.meta['name'] == module:
+                return m
 
-        return mod_paths
+        return None
```
This is the redundant complexity that I mentioned earlier.

We're forced to parse the main `west.yml` file from the `zephyr` module in order to transform `name` into `path` instead of just using `path` in the first place.

I'm also not entirely sure - does this include modules defined outside of the main `west.yml`? I.e. in T2 Star Topology?

MeisterBob replied:
It includes what is configured as `manifest.path`/`manifest.file` in `west config` and everything that is imported from there.

cfriedt reviewed (changes requested):
Some nits, some legitimate concerns (e.g. breaking changes).

Maybe it's time to start adding some test cases (although not fun).

I also should note, tooling changes like this might not be mergable without first going through an architecture review (it was discussed yesterday).

pdgendt commented on scripts/schemas/patch-schema.yml:
```diff
@@ -24,8 +24,7 @@ schema;patch-schema:
 
-        # The path of the module the patch is for, relative to the west workspace
-        # e.g. zephyr, or bootloader/mcuboot
+        # The name of the module the patch is for, e.g. zephyr, mcuboot or hal_nordic
```
Yeah, I did have this concern as well, some of the west extensions do work with module names, rather than paths, but I agree that it doesn't suit this extension much.
You even could end up with some chicken-and-egg problem if a patch modifies the `zephyr/module.yml` file.

MeisterBob replied:
Imho `all_modules = zephyr_module.parse_modules(ZEPHYR_BASE, self.manifest)` isn't too complex to get all modules known to west. By using the name it is clear what to use and one doesn't need to figure out if the path needs to be absolute or relative to current directory, wolrkspace, zephyr, ... . Using the path instead of the name enables you to patch libraries that aren't known to west, but I can't figure out a scenario where this may be required.

Also the number of people affected by the breaking change shouldn't be to many as `west patch` is pretty new.

The module names are taken from `west.yml` not `zephyr/module.yml` (at least from my experience). Modifying `west.yml` with a patch doesn't make much sense because you can overwrite `west.yml` entries of modules you import in your own `west.yml` without fiddling in the other module.

MeisterBob replied:
I evaluated the module name handling a bit more. I have a module defined in my `west.yml`
```yaml
    - name: my_module
      path: modules/lib/my
      revision: ...
      url: ...
      import: true
```
and fired some west commands onto it.
```
west status my_module  -> OK
west compare my_module -> OK
west status my         -> unknown module
west compare my        -> unknown module

west blobs list my_module      -> unknown module
west packages -m my_module pip -> unknown module
west blobs list my             -> OK
west packages -m my pip        -> OK
```
`west list` shows the module as my_module. Build in commands (status, compare, list) handle the
module names different than zephyr's extension commands (blobs, packages).

Adding `name: my1` to the `modules.yml` doesn't change the behaviour for build in commands, but
extension commands now uses `my1` instead of `my`.

Conclusion: Build in commands use the name from `west.yml`, zephyr's extension commands use the name
from `module.yml` when present or falls back to the final part of the module path and ignores the
name from `west.yml`.

I still prefer the use of names, but it might be good to evaluate this behaviour in a seperate issue.

pdgendt replied:
Good analysis, this shows you the difference between "west projects" and "zephyr modules". Built-in west commands (`status`/`compare`) will use "west projects" while most zephyr extensions (`blobs`/`packages`) will use "zephyr modules".

Zephyr modules have a schema for the module definition file (with a `name` for example), needed to describe something about the module other than the current version and more west extensions.

cfriedt replied:
Thanks for looking into this more. There is clearly some inconsistence.

MeisterBob commented on scripts/schemas/patch-schema.yml:
```diff
@@ -11,8 +11,8 @@ schema;patch-schema:
       mapping:
 
-        # The path to the patch file, relative to the root of the module
-        # E.g. zephyr/kernel-pipe-fix-not-k-no-wait-and-ge-min-xfer-bytes.patch
```
Maybe using a patch for zephyr wasn't the best choice as there are a several zephyr folders as you already pointed out. Also the name and path (relative to the west workspace) are the same for zephyr. That may be the major cause to refactor this using names.

cfriedt commented:
I can kick off a build to test this out in our repo in a couple of hours.

cfriedt commented:
> I can kick off a build to test this out in our repo in a couple of hours.

I'm getting a python error when I run this as `west patch apply`

```shell
  File "/Users/cfriedt/tt-zephyr-platforms-work/zephyr/scripts/west_commands/patch.py", line 236, in filter_args
    for i, m in enumerate(args.dst_modules):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
```

cfriedt commented:
I think once the merge conflict and python issue above is fixed, I would probably be OK to switch my `patches.yml` file over to a name-based one, if that greatly simplifies integration with the rest of west or zephyr naming conventions.

I just think there needs to be at least safe commit that I can point my west manifest to first.

mt-gjoerges committed:
scripts: west_commands: patch: add src module argument

The src-module argument can be used to select the module in which west
patch searches for patch definitions. This commit also allows the use of
module names instead of the path.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

cfriedt reviewed (dismissed):
Works for me with the most recent revision. Thanks for pushing forward on this.

MeisterBob commented:
thank you.

Btw: there is no need to update your patch files, the module in `patches.yml` can also be the module name or a relative/absolute path.

pdgendt commented:
> Btw: there is no need to update your patch files, the module in `patches.yml` can also be the module name or a relative/absolute path.

Not sure if this is desired though? My gut is telling me we should do one or the other, supporting both just adds complexity, no?

cfriedt commented:
@pdgendt - probably better for a follow-up PR. As mentioned, there needs to be at least one commit with support for both so that there is an upgrade path without breaking things.

pdgendt commented:
Why though? The extension hasn't been in any Zephyr release, it's less than a month old.

cfriedt commented:
Technically speaking, even an additional commit on top of this PR moving from paths to names would make a transition possible, as long as git history is not destroyed.

It may be easier to do it in another PR though.

Why? To not break things for downstream users, bisectability.

MeisterBob commented:
@cfriedt can you please re-approve
